kv/util/util.go

```diff
@@ -0,0 +1,213 @@
+package util
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/generated/proto/commonpb"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/log"
+)
+
+var errNilStore = errors.New("kv store is nil")
+
+type getValueFn func(kv.Value) (interface{}, error)
+type updateFn func(interface{})
+
+// WatchAndUpdateBool sets up a watch for a bool property.
+func WatchAndUpdateBool(store kv.Store, key string, property *bool, lock sync.Locker,
+	defaultValue bool, logger xlog.Logger) error {
+	updateFn := lockedUpdate(func(i interface{}) { *property = i.(bool) }, lock)
+
+	return watchAndUpdate(store, key, getBool, updateFn, defaultValue, logger)
+}
+
+// WatchAndUpdateFloat64 sets up a watch for an float64 property.
+func WatchAndUpdateFloat64(store kv.Store, key string, property *float64, lock sync.Locker,
+	defaultValue float64, logger xlog.Logger) error {
+	updateFn := lockedUpdate(func(i interface{}) { *property = i.(float64) }, lock)
+
+	return watchAndUpdate(store, key, getFloat64, updateFn, defaultValue, logger)
+}
+
+// WatchAndUpdateInt64 sets up a watch for an int64 property.
+func WatchAndUpdateInt64(store kv.Store, key string, property *int64, lock sync.Locker,
+	defaultValue int64, logger xlog.Logger) error {
+	updateFn := lockedUpdate(func(i interface{}) { *property = i.(int64) }, lock)
+
+	return watchAndUpdate(store, key, getInt64, updateFn, defaultValue, logger)
+}
+
+// WatchAndUpdateTime sets up a watch for a time property.
+func WatchAndUpdateTime(store kv.Store, key string, property *time.Time, lock sync.Locker,
+	defaultValue time.Time, logger xlog.Logger) error {
+	updateFn := lockedUpdate(func(i interface{}) { *property = i.(time.Time) }, lock)
+
+	return watchAndUpdate(store, key, getTime, updateFn, defaultValue, logger)
+}
+
+// BoolFromValue get a bool from kv.Value
+func BoolFromValue(v kv.Value, key string, defaultValue bool, logger xlog.Logger) bool {
+	var res bool
+	updateFn := func(i interface{}) { res = i.(bool) }
+
+	updateWithKV(getBool, updateFn, key, v, defaultValue, logger)
+
+	return res
+}
+
+// Float64FromValue gets an float64 from kv.Value
+func Float64FromValue(v kv.Value, key string, defaultValue float64, logger xlog.Logger) float64 {
+	var res float64
+	updateFn := func(i interface{}) { res = i.(float64) }
+
+	updateWithKV(getFloat64, updateFn, key, v, defaultValue, logger)
+
+	return res
+}
+
+// Int64FromValue gets an int64 from kv.Value
+func Int64FromValue(v kv.Value, key string, defaultValue int64, logger xlog.Logger) int64 {
+	var res int64
+	updateFn := func(i interface{}) { res = i.(int64) }
+
+	updateWithKV(getInt64, updateFn, key, v, defaultValue, logger)
+
+	return res
+}
+
+// StringArrayFromValue gets a string array from kv.Value
+func StringArrayFromValue(v kv.Value, key string, defaultValue []string, logger xlog.Logger) []string {
+	var res []string
+	updateFn := func(i interface{}) { res = i.([]string) }
```

Comment:
it will never happen, the interface of getValueFn and updateFn has to use interface{} since the lack of generic in Go, but within each usage, the type in getValueFn and updateFn should be the same, so the cast is safe